/********************************************************************************
**
**	Copyright (C) 2007-2017 Witold Gantzke & Kirill Lepskiy
**
**	This file is part of the ACF-Solutions Toolkit.
**
**	This file may be used under the terms of the GNU Lesser
**	General Public License version 2.1 as published by the Free Software
**	Foundation and appearing in the file LicenseLGPL.txt included in the
**	packaging of this file.  Please review the following information to
**	ensure the GNU Lesser General Public License version 2.1 requirements
**	will be met: http://www.gnu.org/licenses/old-licenses/lgpl-2.1.html.
**
**	If you are unsure which license is appropriate for your use, please
**	contact us at info@imagingtools.de.
**
** 	See http://www.ilena.org or write info@imagingtools.de for further
** 	information about the ACF.
**
********************************************************************************/


#ifndef iqtcam_CScaleCalibrationEditorComp_included
#define	iqtcam_CScaleCalibrationEditorComp_included


// Qt includes
#include<QtCore/QtGlobal>
#if QT_VERSION >= 0x050000
#include <QtWidgets/QFrame>
#include <QtWidgets/QGroupBox>
#else
#include <QtGui/QFrame>
#include <QtGui/QGroupBox>
#endif

// ACF includes
#include <iqtgui/TDesignerGuiObserverCompBase.h>
#include <iqtgui/IDialog.h>

// ACF-Solutions includes
#include <iipr/IFeaturesProvider.h>

#include <GeneratedFiles/iqtcam/ui_CScaleCalibrationEditorComp.h>


namespace iqtcam
{


class CScaleCalibrationEditorComp: public iqtgui::TDesignerGuiObserverCompBase<Ui::CScaleCalibrationEditorComp, imeas::INumericValue>
{
	Q_OBJECT
public:
	typedef iqtgui::TDesignerGuiObserverCompBase<Ui::CScaleCalibrationEditorComp, imeas::INumericValue> BaseClass;

	I_BEGIN_COMPONENT(CScaleCalibrationEditorComp);
		I_ASSIGN(m_circleProviderCompPtr, "CircleProvider", "External circle provider providing radius used for scale calibration\nThe provider must provide a numeric value of type VTI_RADIUS", false, "CircleProvider");
		I_ASSIGN(m_calibrationDialogCompPtr, "CalibrationDialog", "Dialog supporting parameterization of the circle finder used for image resolution calculation", false, "CalibrationDialog");
		I_REGISTER_INTERFACE(imod::IModelEditor);
	I_END_COMPONENT;

protected:
	// reimplemented (iqtgui::TGuiObserverWrap)
	virtual void UpdateModel() const;
	virtual void UpdateGui(const istd::IChangeable::ChangeSet& changeSet);

	// reimplemented (iqtgui::CGuiComponentBase)
	virtual void OnGuiCreated();

private Q_SLOTS:
	void OnValueChanged(double);
	void on_CalibrateButton_clicked();
	void on_NominalRadiusSpinBox_valueChanged(double d);

private:
	I_REF(iipr::IFeaturesProvider, m_circleProviderCompPtr);
	I_REF(iqtgui::IDialog, m_calibrationDialogCompPtr);
};


} // namespace iqtcam


#endif	// !iqtcam_CScaleCalibrationEditorComp_included


